src/pixel/homogeneous_pixel.rs

```diff
@@ -0,0 +1,125 @@
+use crate::*;
+
+/// A Pixel made up of a compile-time known number of contiguously stored color components and optionally an
+/// alpha component.
+///
+/// Unlike [`HeterogeneousPixel`] the alpha component must be the same type as the color
+/// components.
+///
+/// # Terminology
+///
+/// Component = An element of a pixel, inclusive of alpha. For example, [`Rgba`](crate::Rgba) is a pixel made up
+/// of four components, three color components and one alpha component.
+pub trait HomogeneousPixel:
+    HeterogeneousPixel<ColorComponent = Self::Component, AlphaComponent = Self::Component>
+{
+    /// The component type of the pixel used for both color and alpha components if any.
+    type Component: PixelComponent;
+
+    //TODO if const generic expressions become stable remove this associated type and just use
+    //`[Self::Component; Self::COMPONENT_COUNT]` as the return type.
+    //
+    /// The component array form of `Self`
+    type ComponentArray<R>: IntoIterator<Item = R> + AsSlice<R>;
+
+    /// Converts an owned `Pixel` type to an array of its components.
+    fn component_array(&self) -> Self::ComponentArray<Self::Component>;
```

Comment:
Can this be `impl AsRef<[Component]>`?

Reply:
I'll have a go and see if its possible, I've never used `AsRef` before so this will be fun!

Reply:
I've switched to `impl AsRef<[Self::ColorComponent]> + AsMut<[Self::ColorComponent]> + IntoIterator<Item = Self::ColorComponent>`.

It's a bit long though, perhaps we should put them in a super-trait:
```rust
pub trait ArrayLike<T>: AsRef<[T]> + AsMut<[T]> + IntoIterator<T> {}
```
But that would require another trait which the current versions doesn't.

Reply:
The trait bound is cool, but I actually had in mind `impl AsRef<[C]> for T where T: Pixel<C>`, so instead of `pixel.component_array()`, it'd be enough to call `pixel.as_ref()`.

Reply:
I think this has the same pros/cons as `FromIterator` scenario with the trait bounds vs blanket impls tradeoff. Whichever we decide for that scenario would decide this scenario also I think.